yjit_core.c

```diff
@@ -454,6 +455,18 @@ branch_stub_hit(const uint32_t branch_idx, const uint32_t target_idx, rb_executi
 
     // If this block hasn't yet been compiled
     if (!p_block) {
+        // If the new block will be placed next right after the branch
+        if (cb->write_pos == branch->end_pos) {
+            //fprintf(stderr, "target idx %d will be placed next\n", target_idx);
+            branch->shape = (uint8_t)target_idx;
+
+            // Rewrite the branch with the new, potentially more compact shape
+            cb_set_pos(cb, branch->start_pos);
+            branch->gen_fn(cb, branch->dst_addrs[0], branch->dst_addrs[1], branch->shape);
+            RUBY_ASSERT(cb->write_pos <= branch->end_pos && "can't enlarge branches");
+            branch->end_pos = cb->write_pos;
+        }
```

Comment:
Wdyt about putting this directly before the call to `gen_block_version()`? Also it'd be nice to comment about how the goal with this rewrite is to set cb->write_pos, which `gen_block_version()` uses as a hidden input for the placement of the block.
I would also leave a small note there about how this temporarily corrupts the branch and would need to be repaired in case gen_block_verison fails to compile in extreme circumstances.

Reply:
Sure I'll put it right before `gen_block_version()` and add comments and assertions to make the assumptions clearer.